### ISSUE TITLE
Building wrong-hash FOD with auto-optimise-store on leaves writable files in store

### ISSUE THREAD

**Issue by oxalica** [NixOS/nix#4513]
**Describe the bug**

I have `auto-optimise-store = true` in my `nix.conf`.
When building any fixed output derivation (FOD) with wrong hash, nix prints warning `Suspicious file` on all output files. (For `fetchFromGitHub`, it produces tons of warnings)

**Steps To Reproduce**

1. Set `auto-optimise-store = true` in `nix.conf`.
2. Run `nix build --impure --expr 'with import <nixpkgs> {}; fetchurl { url = "mirror://gnu/hello/hello-2.10.tar.gz"; hash = lib.fakeHash; }'` or the `nix-build -E` variant. (If you already has the path in nix store, `nix-store --delete` it first.)
3. The command fails as expected, but also print an unexpected `Suspicious file` warning.
   ```
   > nix build --impure --expr 'with import <nixpkgs> {}; fetchurl { url = "mirror://gnu/hello/hello-2.10.tar.gz"; hash = lib.fakeHash; }'
   warning: --- Suspicious file --- nix-daemon
   skipping suspicious writable file '/nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz'
   error: --- Error --- nix-daemon
   hash mismatch in fixed-output derivation '/nix/store/xskbrh4581f9zkqwzbnwdwx559d7spxd-hello-2.10.tar.gz.drv':
     specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
        got:    sha256-MeBmE3qWJnbon2nRtlOC3pWn732RS4y5VvQepy4PUWs=
   ```
4. Check the output path in store by `stat`. It has mode 0644 and current mtime,.
   ```
   > stat /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
     File: /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
     Size: 725946          Blocks: 1424       IO Block: 4096   regular file
   Device: 1eh/30d Inode: 2578231     Links: 1
   Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
   Access: 2021-02-03 00:00:10.646606338 +0800
   Modify: 2021-02-03 00:00:10.646606338 +0800
   Change: 2021-02-03 00:00:10.648606398 +0800
    Birth: 2021-02-03 00:00:10.646606338 +0800
   ```

**Expected behavior**

- The build command failed with hash mismatch without any other warnings.
- The output path should have mode 0444 and mtime 1.

**`nix-env --version` output**
`nix-env (Nix) 2.4pre20201205_a5d85d0`

**Additional context**

When `auto-optimise-store` is disabled by running `nix build ... --option auto-optimise-store false`, it works as expected. There is no warning, and the output path has correct mode and mtime.

**Comment by regnat** [comment]
This might be fixed by https://github.com/NixOS/nix/pull/4776 (at least it’s worth checking out)

**Comment by oxalica** [comment]
Confirmed it's fixed in b68a85df0afd89b40c93a9441c2a29ff6c4ba3d1
